pymeasure/instruments/anritsu/AnritsuMS9740A.py

```diff
@@ -0,0 +1,140 @@
+from time import sleep
+import numpy as np
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.anritsu import AnritsuMS9710C
+from pymeasure.instruments.validators import (
+    strict_discrete_set,
+    truncated_discrete_set,
+    truncated_range,
+    joined_validators
+)
+import re
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+# Analysis Results with Units, ie -24.5DBM -> (-24.5, 'DBM')
+r_value_units = re.compile("([-\d]*\.\d*)(.*)")
+
+# Join validators to allow for special sets of characters
+truncated_range_or_off = joined_validators(strict_discrete_set, truncated_range)
+
+
+def _int_or_neg_one(v):
+    try:
+        return int(v)
+    except ValueError:
+        return -1
+
+
+def _parse_trace_peak(vals):
+    """Parse the returned value from a trace peak query."""
+    l, p = vals
+    res = [l]
+    m = r_value_units.match(p)
+    if m is not None:
+        data = list(m.groups())
+        data[0] = float(data[0])
+        res.extend(data)
+    else:
+        res.append(float(p))
+    return res
+
+
+class AnritsuMS9740A(AnritsuMS9710C):
+    """Anritsu MS9740A Optical Spectrum Analyzer."""
+    
+    #############
+    #  Mappings #
+    #############
+    
+    ONOFF = ["ON", "OFF"]
+    ONOFF_MAPPING = {True: 'ON', False: 'OFF', 1: 'ON', 0: 'OFF'}
+
+    ###########
+    #  Modes  #
+    ###########
+
+    measure_mode = Instrument.measurement(
+        "MOD?", "Returns the current Measure Mode the OSA is in.",
+        values={None: 0, "SINGLE": 1.0, "REPEAT": 2.0, "POWER": 3.0},
+        map_values=True
+    )
+
+    ####################################
+    # Spectrum Parameters - Wavelength #
+    ####################################
+    
+    resolution = Instrument.control(
+        "RES?", "RES %s", "Resolution (nm)",
+        validator=truncated_discrete_set,
+        values=[0.03, 0.05, 0.07, 0.1, 0.2, 0.5, 1.0],
+    )
+
+    resolution_vbw = Instrument.control(
+        "VBW?", "VBW %s", "Video Bandwidth Resolution",
+        validator=strict_discrete_set,
+        values=["1MHz", "100kHz", "10kHz", "2kHz", "1kHz", "200Hz", "100Hz", "10Hz"]
+    )
+
+    average_sweep = Instrument.control(
+        "AVS?", "AVS %d",
+        "Number of averages to make on a sweep (1-1000)",
+        validator=truncated_range, values=[1, 1000]
+    )
+
+    sampling_points = Instrument.control(
+        "MPT?", "MPT %d", "Number of sampling points",
+        validator=truncated_discrete_set,
+        values=[51, 101, 251, 501, 1001, 2001, 5001, 10001, 20001, 50001],
+        get_process=lambda v: int(v)
+    )
+
+    ##########################
+    #  Data Memory Commands  #
+    ##########################
+
+    data_memory_select = Instrument.control(
+        "TTP?", "TTP %s",
+        "Memory Data Select.",
+        validator=strict_discrete_set,
+        values=["A", "B", "C", "D", "E", "F", "G", "H", "I", "J"]
+    )
+
+    def __init__(self, adapter, **kwargs):

```

Comment:
Can you please place the __init__ at the beginning of the class? It's not really homogeneous over the codebase yet, but i try to do that for new instruments.

Reply:
I have moved it, however I am actually having isses with the code. I have only just got access to the instrument and when i try to call my new class I get an error in the new __init__ statement which I was hoping you could help with? When running my class locally as a stand alone (ie no parents) the __init__works fine, however when adding (AnritsuMS9710C) in the Class line, I get a "TypeError: __init_ takes 2 positional arguments but 3 were given". 
As far as I can tell it was receiving 3 arguments (namely adapter, a string and **kwargs) in the parent class and that works so im unsure what the issue is. I guess its something to do with adapter (which I have no idea what it is or where it is defined/imported)

Reply:
Changing the definition of the parent class in anritsuMS9710C.py to read
```
    def __init__(self, adapter, name="Anritsu MS9710C Optical Spectrum Analyzer", **kwargs):
        """Constructor."""
        self.analysis_mode = None
        super(AnritsuMS9710C, self).__init__(adapter, name=name, **kwargs)
```

and the super call here in your daughter class init to
```
super(AnritsuMS9740A, self).__init__(adapter, name="Anritsu MS9740A Optical Spectrum Analyzer", **kwargs)
```
should do the trick, and when you construct an instance, it should have the correct ` myinstr.name` attribute.

Previously, the parent class did not expect to have its name overridden with a constructor argument, so that caused the problems you were seeing.